Refactor some registry client and auth code

jlhawn opened the pull request:
Adds NewClient() method to registry.Endpoint objects to make it much
simpler to create and use a http client for the registry package.

Switched registry.Session to use a single http client for reuse and
connection pooling.

Refactored the newClient() logic which configured custom server and
client certs for registry endpoints.

Updated registry package unit tests to use two different HTTP test servers,
one with a valid certificate configured by the test and another without.

Added checks to ensure that the registry session DOES NOT send auth
credentials over an insecure connection.

Docker-DCO-1.1-Signed-off-by: Josh Hawn josh.hawn@docker.com (github: jlhawn)

Commit:
Refactor some registry client and auth code

Adds NewClient() method to registry.Endpoint objects to make it much
simpler to create and use a http client for the registry package.

Switched registry.Session to use a single http client for reuse and
connection pooling.

Refactored the newClient() logic which configured custom server and
client certs for registry endpoints.

Updated registry package unit tests to use two different HTTP test servers,
one with a valid certificate configured by the test and another without.

Added checks to ensure that the registry session DOES NOT send auth
credentials over an insecure connection.

Custom CA certs and client certs are now loaded automatically for every
registry session request and reloaded automatically when there are any
changes to the custom certs directory. NOTE: This means that if you are
using any custom certs, the registry client will no longer use system
root certs unless you also add those manually.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
ping @tiborvass @dmcgowan

jlhawn commented on registry/endpoint.go:
```diff
@@ -245,7 +268,7 @@ func (e *Endpoint) pingV2() (RegistryInfo, error) {
 		// It would seem that no authentication/authorization is required.
 		// So we don't need to parse/add any authorization schemes.
-		return RegistryInfo{Standalone: true}, nil
+		return RegistryInfo{}, nil
```
I just removed this because a v2 registry no longer has the concept of a "standalone" registry.

tiborvass commented:
I also tried some refactoring at https://github.com/tiborvass/docker/compare/docker:master...fix-login-secure-registry, and the biggest problem is that you need the request object to check if you are able to send creds securely (https://github.com/tiborvass/docker/compare/docker:master...fix-login-secure-registry#diff-7025b8c38be826dd40b8f9f14310b7c1R75)

Another thing to keep in mind, is that we actually do want to create a client *almost* every single time, so that new certs get caught in /etc/docker. An improvement would be to reuse the same client for each registry command exposed by the API. A major improvement would be to use http://godoc.org/github.com/howeyc/fsnotify and watch the /etc/docker/certs.d subtree.

stevvooe commented on registry/endpoint.go:
```diff
@@ -54,12 +54,12 @@ func NewEndpoint(index *IndexInfo) (*Endpoint, error) {
 	// Try HTTPS ping to registry
 	endpoint.URL.Scheme = "https"
-	if _, err := endpoint.Ping(); err != nil {
+	if endpoint.RegInfo, err = endpoint.Ping(); err != nil {
```
Could this unintentionally clear out `endpoint.RegInfo on error? Or is that behavior okay?

jlhawn replied:
Should be okay since a ping attempt is made again after this, and if that one fails then the endpoint is no good anyway.

stevvooe replied:
Cheers!

jlhawn commented:
@tiborvass I've tried to address the points you brought up in the latest update.

> the biggest problem is that you need the request object to check if you are able to send creds securely 

[This change](https://github.com/docker/docker/pull/10265/files#diff-e2274d1b865692b0cac54b3a57a6785eR48) makes any registry session instance use a custom http client wrapper which knows whether `InsecureSkipVerify` is true or not and checks for an `Authorization` header before making any requests. If the request is found to be insecure, then it refuses to make the request with any non-empty value for the `Authorization` header.

> Another thing to keep in mind, is that we actually do want to create a client almost every single time, so that new certs get caught in /etc/docker. An improvement would be to reuse the same client for each registry command exposed by the API.

This change makes it so that each call to `registry.NewSession()` gets a new client instance that is used for the remainder of the session.

> A major improvement would be to use http://godoc.org/github.com/howeyc/fsnotify and watch the /etc/docker/certs.d subtree.

It looks like we already vendor [go-fsnotify](https://github.com/go-fsnotify/fsnotify) which is awesome! I've updated this patch to use this for setting up a global CA Pool and set of client certificates. I've also added some unit tests for this and made some other improvements to the `registry` package unit tests in general.

dkjer commented on registry/client.go:
```diff
@@ -0,0 +1,278 @@
+// Auth headers are never sent over an insecure channel.
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader != "" && !(c.isSecure && req.URL.Scheme == "https") {
```
Does this mean users may accidentally send Auth headers over an insecure channel is they both set --insecure-registry and use http:// ?

jlhawn replied:
perhaps there are too many boolean operators on this line making it too complex :(

the condition in parentheses (`c.isSecure && req.URL.Scheme == "https"`) would mean, in plain english, that the request will use "real" secure TLS for the connection not `InsecureSkipVerify`. If this is not the case then the client will refuse to make a request with any Authorization header.

dkjer replied:
Ah, right.  So using Auth headers over https with a self-signed cert will no longer work with --insecure-registry?

stevvooe commented on registry/client.go:
```diff
@@ -0,0 +1,278 @@
+)
+
+var (
+	caPool       *x509.CertPool
```
Please factor this into a structure so that their relation is clear.

stevvooe commented on registry/client.go:
```diff
@@ -0,0 +1,278 @@
+	return
+}
+
+func watchCertDirs() (certDirs []string) {
```
Shouldn't the daemon require a restart or hup signal to reload certificates? It seems like a security hole to have this changing whenever the file system changes.

crosbymichael commented:
Since this has refactoring and new features from your PR description `Added checks to ensure that the registry session DOES NOT send auth
credentials over an insecure connection.`  I would suggest breaking this PR up and if you want to do the refactor work, do that first and get it merged in before changing the functionality.  

Please break this PR up and reopen when you are ready for review if you still want to add these changes.  This helps the diff to be smaller, making it easier to review, and we can focus on the added feature or the refactoring and not both at the same time.

jlhawn commented:
@crosbymichael thanks, I think I can add the registry client refactor, certs dir watching, and unsafe auth header each as separate PRs, in that order.
